### PR TITLE
Issue 65 skip resolve

### PR DESCRIPTION
**Pull request by adithyabsk**
### Related Issue
Fixes issue #65 

### Description
* Add feature to SmartTransformer to allow one to skip the resolution stage.
* Re-organize utilities and clean SmartTransformer to be more lean.
* Patch pandas_wrapper to work more elegantly with _Empty (which will be deprecated)

### CODE REVIEW

**Comment by adithyabsk** [comment]
TODO: Increase test coverage before merge

File: foreshadow/transformers/base.py [hunk 1]
```diff
@@ -440,91 +483,56 @@ def set_params(self, **params):
...
 
+        # Calls to override auto set the transformer instance
         self.override = params.pop("override", self.override)
-        if self.override is not None:
-            self._transformer = self._resolve(self.override)(**self.kwargs)
 
-        if self._transformer is not None:
+        if self.transformer is not None:
             valid_params = {
                 k.partition("__")[2]: v
```
Comment (cchoquette) [inline foreshadow/transformers/base.py]: why is this needed?

File: foreshadow/transformers/base.py [hunk 2]
```diff
@@ -440,91 +483,56 @@ def set_params(self, **params):
...
-        # Propagate name and keep_columns attributes to transformer
-        self._transformer.name = self.name
-        self._transformer.keep_columns = self.keep_columns
+        # If override is passed in or set, all types of resolves are turned
+        # off.
+        # Otherwise, force_reresolve will always resolve on each fit.
+
+        # If force_reresolve is set, always re-resolve
+        if self.force_reresolve:
+            self.should_resolve = True
```
Comment (cchoquette) [inline foreshadow/transformers/base.py]: why do we need should_resolve? it seems as though it should be the standard behaviour.

File: foreshadow/utils/validation.py [hunk 3]
```diff
@@ -109,22 +114,77 @@ def check_transformer_imports(printout=True):
...
+                    class_name, ", ".join(sources.keys())
+                )
+            )
+
+    return getattr(module, class_name)
+
 
-            traceback.print_exception(type, value, tb)
-            pdb.post_mortem(tb)
+def is_transformer(value, method="isinstance"):
```
Comment (cchoquette) [inline foreshadow/utils/validation.py]: when do we want to switch it?

File: foreshadow/tests/test_transformers/test_transformers.py [hunk 4]
```diff
@@ -266,26 +267,15 @@ def test_transformer_pipeline():
...
 def test_smarttransformer_notimplemented():
-    import pandas as pd
-
     from foreshadow.transformers.base import SmartTransformer
 
-    boston_path = get_file_path("test_data", "boston_housing.csv")
-
-    df = pd.read_csv(boston_path)
-
     class TestSmartTransformer(SmartTransformer):
```
[the remark this answers is not in the document]
Comment (adithyabsk) [reply foreshadow/tests/test_transformers/test_transformers.py]: Add test for instantiate SmartTransformer

File: foreshadow/tests/test_transformers/test_transformers.py [hunk 5]
```diff
@@ -325,35 +316,55 @@ class InvalidClass:
...
-    )
+        "is neither an sklearn Pipeline, FeatureUnion, a "
+        "wrapped foreshadow transformer, nor None."
+    ) in str(e.value)
 
 
-def test_smarttransformer_function():
-    import pandas as pd
+@pytest.fixture()
+def smart_child():
```
[the remark this answers is not in the document]
Comment (adithyabsk) [reply foreshadow/tests/test_transformers/test_transformers.py]: Mock the return value of the SmartTransformer so that the fixture can be used everywhere.

File: foreshadow/tests/test_transformers/test_transformers.py [hunk 6]
```diff
@@ -266,26 +267,15 @@ def test_transformer_pipeline():
...
+        TestSmartTransformer()
 
-    assert (
-        str(e.value) == "WrappedTransformer _get_transformer was not "
-        "implimented."
-    )
+    assert "Can't instantiate abstract class" in str(e.value)
 
 
 def test_smarttransformer_attributeerror():
```
[the remark this answers is not in the document]
Comment (adithyabsk) [reply foreshadow/tests/test_transformers/test_transformers.py]: At least from the school of thought I come from, each unit test should be testing a separate functionality. I can change if it you feel really strongly about this, though.

File: foreshadow/tests/test_transformers/test_transformers.py [hunk 7]
```diff
@@ -399,85 +415,76 @@ class TestSmartTransformer(SmartTransformer):
...
-        pass
+    Args:
+        smart_child: A subclass of SmartTransformer.
 
-    smart = TestSmartTransformer()
+    """
+    smart = smart_child()
     smart.set_params()
 
     assert smart._transformer is None
```
[the remark this answers is not in the document]
Comment (adithyabsk) [reply foreshadow/tests/test_transformers/test_transformers.py]: Good catch, will fix.

File: foreshadow/tests/test_transformers/test_transformers.py [hunk 8]
```diff
@@ -497,17 +504,24 @@ class TestSmartTransformer(SmartTransformer):
...
 
-    smart = TestSmartTransformer()
+    smart = smart_child()
     smart.fit([])
 
-    assert smart.inverse_transform([1, 2, 10]).size == 0
+    with pytest.raises(InverseUnavailable) as e:
+        smart.inverse_transform([1, 2, 10])
+
+    assert (
```
[the remark this answers is not in the document]
Comment (adithyabsk) [reply foreshadow/tests/test_transformers/test_transformers.py]: I usually keep the string so that the test fails when the string changes so I know to update the docs.

File: foreshadow/transformers/base.py [hunk 9]
```diff
@@ -1,17 +1,20 @@
...
     FeatureUnion,
+    Pipeline,
     _fit_one_transformer,
     _fit_transform_one,
     _transform_one,
 )
 
-from foreshadow.transformers.transformers import _Empty
-from foreshadow.utils import check_df
+from foreshadow.transformers.transformers import make_pandas_transformer
```
[the remark this answers is not in the document]
Comment (adithyabsk) [reply foreshadow/transformers/base.py]: Resolved in next PR

File: foreshadow/transformers/base.py [hunk 10]
```diff
@@ -316,7 +319,9 @@ def fit_transform(self, X, y=None, **fit_params):
         return Xs
 
 
-class SmartTransformer(BaseEstimator, TransformerMixin):
+# TODO: Remove once _Empty is removed when DataCleaner is implemented
+@make_pandas_transformer
+class SmartTransformer(BaseEstimator, TransformerMixin, metaclass=ABCMeta):
```
[the remark this answers is not in the document]
Comment (adithyabsk) [reply foreshadow/transformers/base.py]: This was changed so that "get_transformer" would error at class definition time rather than runtime.

File: foreshadow/transformers/base.py [hunk 11]
```diff
@@ -326,51 +331,128 @@ class SmartTransformer(BaseEstimator, TransformerMixin):
...
-    implementing class that returns an sklearn transformer object to be used.
+    implementing class that returns an scikit-learn transformer object to be
+    used.
 
     Used and implements itself identically to a transformer.
 
     Attributes:
-        override: An sklearn transformer that can be optionally provided to
-            override internals logic.
+        override: An scikit-learn transformer that can be optionally provided
```
[the remark this answers is not in the document]
Comment (adithyabsk) [reply foreshadow/transformers/base.py]: Will fix

File: foreshadow/transformers/base.py [hunk 12]
```diff
@@ -326,51 +331,128 @@ class SmartTransformer(BaseEstimator, TransformerMixin):
...
 
         """
-        if self._transformer is None:
-            raise ValueError("Smart Transformer not Fit")
-        else:
-            return self._transformer
+        # Check transformer type
+        valid_tranformer = (
+            is_transformer(value)
+            and hasattr(value, "name")  # check wrapping
```
[the remark this answers is not in the document]
Comment (adithyabsk) [reply foreshadow/transformers/base.py]: Will fix.

File: foreshadow/transformers/base.py (hunk 1 above)
Comment (adithyabsk) [reply foreshadow/transformers/base.py]: I didn't write this piece of code, it's a holdover from the initial version I left unchanged, I can dig into it later.

File: foreshadow/transformers/base.py (hunk 2 above)
Comment (adithyabsk) [reply foreshadow/transformers/base.py]: Will fix. will have singular statement checking whether force resolve is set, and will update should resolve as a result of that.

File: foreshadow/utils/validation.py (hunk 3 above)
Comment (adithyabsk) [reply foreshadow/utils/validation.py]: Will resolve later.

File: foreshadow/transformers/transformers.py [hunk 13]
```diff
@@ -388,6 +388,15 @@ def fit_transform(self, X, *args, **kwargs):
                     out = _keep_columns_process(out, df, name)
             return out
 
+        def is_wrapped(self):
```
Comment (cchoquette) [inline foreshadow/transformers/transformers.py]: I just realized, we shouldn't do a method because it will always exist and return True. When it doesnt, which is what we are trying to check, it will error out with an AttributeError which is not that graceful - likely better to use a utility function of foreshadow.utils to check and return True or False.

Comment (adithyabsk) [reply foreshadow/transformers/transformers.py]: Will patch

File: foreshadow/utils/validation.py [hunk 14]
```diff
@@ -188,3 +188,16 @@ def is_transformer(value, method="isinstance"):
...
+    """Check if a transformer is wrapped.
+
+    Args:
+        transformer: A transformer instance
+
+    Returns:
+        bool: Whether or not the transformer is wrapped.
+
+    """
+    return "make_pandas_transformer" in repr(transformer.__class__)
```
Comment (cchoquette) [inline foreshadow/utils/validation.py]: Better to check MRO but let's fix that later.

**Comment by adithyabsk** [comment]
@cchoquette Good point, the MRO will be the same as a repr. To future proof this implementation, I switched it up to add an attribute that is checked for.
